### PR TITLE
Feature/user

### PR DESCRIPTION
**Pull request by Lean0411**
## 進度
This pull request includes several changes to add test dependencies and scripts, create test files for user login and registration, and refactor server initialization into a separate file.

### Testing Setup:

- package.json: Added jest as the test script and included supertest as a development dependency. 

- src/tests/user.login.test.js: Tests for successful login, incorrect password, and non-existent user scenarios.

- src/tests/user.register.test.js: Tests for successful registration, existing user, and missing username/password scenarios.

### Codebase Refactoring:

* [`src/app.js`](diffhunk://#diff-c72a907ac323cd2f334ed0e2bd07d15ab62581c4753660c8a0d1c681b30be4b6L41-R42): Removed server initialization code to modularize the application setup.

* [`src/server.js`](diffhunk://#diff-79f4c7cdc0b026f6e556196f5911a449157807cffdb3f9d33f0b5404b4ffc2f3R1-R7): Added a new file to handle server initialization, ensuring a cleaner separation of concerns.

### Code Improvements:

* [`src/controllers/userController.js`](diffhunk://#diff-6210ace6be99d7adc7dc8298a158709fef7ba29f17e9c3bd35f1f8fab8d9c8d1R8-R25): Added validation to check for empty username or password during user registration.
- Add login unit test
- Add register unit test
- Separate server logic into server.js

### 測試出現錯誤：

原因每次都要啟動 port 3000，這樣測試完 port 都不會關，是不是可以用這個分離解決這個問題

```bash
  ● Test suite failed to run

    listen EADDRINUSE: address already in use :::3000

      41 | // start server in env.PORT or 3000
      42 | const PORT = process.env.PORT || 3000;
    > 43 | app.listen(PORT, () => {
         |     ^
      44 |     console.log(`server is running ${PORT}`);
      45 | });
      46 |

      at Function.listen (node_modules/express/lib/application.js:635:24)
      at Object.listen (src/app.js:43:5)
      at Object.require (src/__tests__/user.login.test.js:4:13)

Test Suites: 1 failed, 1 passed, 2 total
Tests:       6 passed, 6 total
Snapshots:   0 total
Time:        0.991 s, estimated 1 s
Ran all test suites.
Jest did not exit one second after the test run has completed.

'This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```

### 參考資料：

https://dev.to/superiqbal7/separating-app-and-server-in-express-why-it-matters-and-how-it-benefits-your-application-149d

### CODE REVIEW

File: src/__tests__/user.login.test.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+const User = require('../models/User');
+
+beforeAll(async () => {
+    await mongoose.connect(process.env.MONGODB_URI, {
+        useNewUrlParser: true,
+        useUnifiedTopology: true,
+    });
+
+    // clear users set
+    await User.deleteMany({});
```
Comment (az879593) [inline src/__tests__/user.login.test.js:15]: use drop

File: src/__tests__/user.login.test.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    
+    await new User({
+        username: 'testuser',
+        password: 'testpassword',
+    }).save();
+});
+
+afterAll(async () => {
+    await mongoose.connection.db.dropDatabase();
```
Comment (az879593) [inline src/__tests__/user.login.test.js:24]: only need to drop user collection

File: src/__tests__/user.register.test.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    // connect mongoDB
+    await mongoose.connect(process.env.MONGODB_URI, {
+        useNewUrlParser: true,
+        useUnifiedTopology: true,
+    });
+});
+
+afterAll(async () => {
+    // clear mongoDB
+    await mongoose.connection.db.dropDatabase();
```
Comment (az879593) [inline src/__tests__/user.register.test.js:19]: only need to drop user collection

File: src/__tests__/user.register.test.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+describe('User Register Test', () => {
+    // register success
+    test('Success user register', async () => {
+        const response = await request(app).post('/api/user/register').send({
+            username: 'Lean0411',
+            password: '910411',
+        });
+        expect(response.status).toBe(201);
+        expect(response.body.message).toBe('register success');
```
Comment (az879593) [inline src/__tests__/user.register.test.js:32]: 這裡不應該用 res.body.message 去設 assert
你在這應該做的是去資料庫把這筆資料抓出來
看能不抓到你剛剛 register 的這筆資料
最簡單就是 register 完後再用剛 register 的帳號去 login

**Review by az879593 — CHANGES_REQUESTED** [review]

File: src/__tests__/user.login.test.js (hunk 2 above)
Comment (Lean0411) [reply src/__tests__/user.login.test.js:24]: 所以不需要`await mongoose.connection.close();嗎？
為什麼？

Comment (Lean0411) [reply src/__tests__/user.login.test.js:24]: 如果不關閉`await mongoose.connection.close()`
unit test 會出現錯誤

## 測試結果與錯誤訊息

```shell
Test Suites: 3 passed, 3 total
Tests:       12 passed, 12 total
Snapshots:   0 total
Time:        0.936 s, estimated 1 s
Ran all test suites.
Jest did not exit one second after the test run has completed.

This usually means that there are asynchronous operations that weren't stopped in your tests.
Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.
```
截圖 2024-10-26 下午5 14 06

**Review by az879593 — APPROVED** [review]

**Pull request merged**